### ISSUE TITLE
NuGet sometimes treats a remote source as a local file

### ISSUE THREAD

**Issue by mesbahamin** [NuGet/Home#6024]
# NuGet sometimes treats a remote source as a local file

## Summary

- I have a project with a custom remote NuGet source called `myorganization-Combined`.
- I execute `nuget restore` on the project's parent solution, specifying that
  the dependencies should be obtained from that custom source.
- This happens successfully for the other projects in the solution.
- For this particular project, however, I get the following exception: `Failed
  to retrieve information about 'somePackage' from remote source
  'C:\dev\myorganization\mySolution\src\myorganization-Combined'.`
- `myorganization-Combined` is the name of a remote NuGet source, yet NuGet.exe is looking
  for `C:\dev\myorganization\mySolution\src\myorganization-Combined`, a file that doesn't exist.

## Details about Problem

| NuGet product used | NuGet version | dotnet.exe --version | OS version              |
| ------------------ | ------------- | -------------------- | ----------------------- |
| NuGet.exe          | 4.3.0.4406    | 2.0.2                | win10 v1703 (15063.632) |

### Other Context

```
$ dotnet --info
.NET Command Line Tools (2.0.2)

Product Information:
 Version:            2.0.2
 Commit SHA-1 hash:  a04b4bf512

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15063
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.0.2\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.0
  Build    : e8b8861ac7faf042c87a5c2f9f2d04c98b69f28d
```

```
$ nuget sources
Registered Sources:

  1.  nuget.org [Disabled]
      https://api.nuget.org/v3/index.json
  2.  myorganization-Combined [Enabled]
      https://myorganization.jfrog.io/myorganization/api/nuget/nuget-combined
  3.  Microsoft Visual Studio Offline Packages [Enabled]
      C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\
```

## Repro steps

1. Disable nuget.org as a source.

2. Add and enable a source called "myorganization-Combined", which is a remote source hosted
   on jfrog.io.

3. Have a solution called mySolution.sln with multiple projects with nuget
   dependencies hosted on myorganization-Combined, including one project,
   myorganization.myProject.Web, targeting .NET Framework 4.7 

4. Successfully build the mySolution solution.

5. Successfully run and test myProject.

6. Run `nuget restore mySolution.sln -Verbosity detailed -Source myorganization-Combined`.

7. Expect success. Note that NuGet successfully acquires and restores packages
   for most projects.

8. Encounter the exception in the log below for myorganization.myProject.Web, where NuGet
   seems to look for a local file with the name `myorganization-Combined`.

### Verbose Logs

- Private information has been removed from this log.

```
NuGet Version: 4.3.0.4406
MSBuild auto-detection: using msbuild version '15.4.8.50001' from 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\bin'. Use option -MSBuildVersion to force nuget to use a specific version of MSBuild.
MSBuild P2P timeout [ms]: 120000
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\bin\msbuild.exe "C:\Users\myusername\AppData\Local\Temp\NuGetScratch\cl53ub1d.ehs.nugetinputs.targets" /t:GenerateRestoreGraphFile /nologo /nr:false /v:q /p:NuGetRestoreTargets="C:\Users\myusername\AppData\Local\Temp\NuGetScratch\5vp2p2ri.qkw.nugetrestore.targets" /p:RestoreTaskAssemblyFile="C:\ProgramData\chocolatey\lib\NuGet.CommandLine\tools\NuGet.exe" /p:RestoreSources="myorganization-Combined;"  /p:RestoreSolutionDirectory="C:\dev\myorganization\mySolution\src"

Restoring NuGet package Microsoft.IdentityModel.Tokens.5.1.3.
Restoring NuGet package Microsoft.IdentityModel.Logging.1.1.3.

[ Redacted: A bunch of other nuget packages for various projects are successfully restored and added. ]

Restoring packages for C:\dev\myorganization\mySolution\src\myorganization.myProject.Web\myProject\myorganization.myProject.Web.csproj...
Restoring packages for .NETFramework,Version=v4.7...
Restoring packages for .NETFramework,Version=v4.6.2...
Restoring packages for .NETFramework,Version=v4.7...
Restoring packages for .NETFramework,Version=v4.7...
Failed to retrieve information about 'somePackage' from remote source 'C:\dev\myorganization\mySolution\src\myorganization-Combined'.
NuGet.Protocol.Core.Types.FatalProtocolException: Failed to retrieve information about 'somePackage' from remote source 'C:\dev\myorganization\mySolution\src\myorganization-Combined'.
   at NuGet.Protocol.LocalV3FindPackageByIdResource.GetVersionsCore(String id, ILogger logger)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at NuGet.Protocol.LocalV3FindPackageByIdResource.GetVersions(String id, SourceCacheContext cacheContext, ILogger logger)
   at NuGet.Protocol.LocalV3FindPackageByIdResource.GetAllVersionsAsync(String id, SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
   at NuGet.Commands.SourceRepositoryDependencyProvider.<GetAllVersionsAsync>d__24.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.SourceRepositoryDependencyProvider.<FindLibraryCoreAsync>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.SourceRepositoryDependencyProvider.<>c__DisplayClass16_0.<<FindLibraryAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.SourceRepositoryDependencyProvider.<FindLibraryAsync>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.DependencyResolver.ResolverUtility.<>c__DisplayClass7_1.<<FindLibraryFromSourcesAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.DependencyResolver.ResolverUtility.<FindLibraryFromSourcesAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.DependencyResolver.ResolverUtility.<FindLibraryByVersionAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.DependencyResolver.ResolverUtility.<FindLibraryMatchAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.DependencyResolver.ResolverUtility.<FindLibraryEntryAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.DependencyResolver.RemoteDependencyWalker.<CreateGraphNode>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.DependencyResolver.RemoteDependencyWalker.<CreateGraphNode>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.DependencyResolver.RemoteDependencyWalker.<CreateGraphNode>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.ProjectRestoreCommand.<WalkDependenciesAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.ProjectRestoreCommand.<TryRestore>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreCommand.<ExecuteRestoreAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreCommand.<ExecuteAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreRunner.<ExecuteAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreRunner.<ExecuteAndCommitAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreRunner.<CompleteTaskAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreRunner.<RunAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreRunner.<RunAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Commands.RestoreRunner.<RunAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.CommandLine.RestoreCommand.<ExecuteCommandAsync>d__34.MoveNext()
```

**Comment by mesbahamin** [comment]
I think this is actually a duplicate of #5849 

Contributing factors to the great deal of confusion I experienced:
- The `-Source` parameter is treated inconsistently between nuget commands (mentioned in #5767). With `nuget install` I can pass it the name of one of my configured sources, but with `nuget restore` I can only pass it a URL.
- The [documentation](https://docs.microsoft.com/en-us/nuget/tools/nuget-exe-cli-reference#restore) doesn't make clear this inconsistency, or even what form the `-Source` argument is supposed to take (URL? File path? Source name?):
    > Source - Specifies list of package sources to use for the restore.
- The failure message in the log I included above is a bit misleading. Fortunately, this appears to have been improved as of 4.6.0.0.
